### DISCUSSION TITLE
Диего и ожидание

### DISCUSSION THREAD

**Discussion by dosinabox** [dosinabox/g2nr_unofficial_update#388] [Ideas]
Если попросить Диего подождать около шахты Сильвестро, то он не дойдет до точки SPAWN_OW_SCAVENGER_01_DEMONT5, контролирующей его местоположение, из-за чего будет отказываться ходить с ГГ. Исправляется сном или уходом за зону видимости, после чего его положение сбрасывается на правильное. Также его можно ударить и привести к этой точке.

Также нужно проверить остальные точки:
```
func int C_DiegoTooFar(var int tolerance)
{
	if(CurrentLevel != OLDWORLD_ZEN)
	{
		return FALSE;
	}
	else if(Npc_GetDistToWP(hero,"SPAWN_OW_SHADOWBEAST_10_01") < (4000 + tolerance))
	{
		return LOC_ANGAR;
	}
	else if(Npc_GetDistToWP(hero,"OW_ICEREGION_ENTRANCE_01") < (3000 + tolerance))
	{
		return LOC_ICE;
	}
	else if(Npc_GetDistToWP(hero,"OW_PATH_046") < (1800 + tolerance))
	{
		return LOC_SWAMP;
	}
	else if(Npc_GetDistToWP(hero,"CASTLE_5") < (1500 + tolerance))
	{
		return LOC_FIRE;
	}
	else if((Npc_GetDistToWP(hero,"MT16") < (6500 + tolerance)) || (Npc_GetDistToWP(hero,"MT13") < (5500 + tolerance)))
	{
		return LOC_LAKE;
	}
	else if((Npc_GetDistToWP(hero,"DT_MAINGATE") < (500 + tolerance)) || (Wld_GetPlayerPortalGuild() == GIL_DMT) || ((Npc_GetDistToWP(hero,"DT_E3_03") < 1000) && (Npc_GetHeightToNpc(self,hero) > 1000)))
	{
		return LOC_XARDAS;
	}
	else if((Npc_GetDistToWP(hero,"OW_NEWMINE_03") < (2500 + tolerance)) || (Npc_GetDistToWP(hero,"OW_NEWMINE_03_B") < (2500 + tolerance)))
	{
		return LOC_FAJETHMINE;
	}
	else if(Npc_GetDistToWP(hero,"OW_MINE3_OUT") < (1200 + tolerance))
	{
		return LOC_SILVESTROMINE;
	}
	else if(Npc_GetDistToWP(hero,"OW_PATH_266") < (3000 + tolerance))
	{
		return LOC_GRIMESMINE;
	}
	else if((Npc_GetDistToWP(hero,"OC_CENTER_02") < (12000 + tolerance)) || (Npc_GetDistToWP(hero,"OC_CENTER_05") < (12000 + tolerance)))
	{
		return LOC_BURG;
	}
	else if((Npc_GetDistToWP(hero,"OW_ORCBARRIER_08") < (5000 + tolerance)) || (Npc_GetDistToWP(hero,"PATH_OC_PSI_01") < (5000 + tolerance)))
	{
		return LOC_ORCBARRIER;
	}
	else if(Npc_GetDistToWP(hero,"OW_PATH_BLOODFLY01_SPAWN01") < (5000 + tolerance))
	{
		return LOC_ORCBARRIER_FAR;
	}
	else if(Npc_GetDistToWP(hero,"PATH_TO_PLATEAU04_SMALLPATH") < (1000 + tolerance))
	{
		return LOC_ROCK;
	};
	return FALSE;
};
```

**Comment by Kor-Angar** [comment]
и Ларес. он тоже у меня часто виснет возле моста по дороге к пирамиде.

**Comment by dosinabox** [comment]
Сохраненки нужны. С Ларесом до моста и с Диего около его пещеры.

**Comment by Kor-Angar** [comment]
нету сейва

**Comment by Kor-Angar** [comment]
> Исправляется сном или уходом за зону видимости, после чего его положение сбрасывается на правильное. 

не работает. вышел из мира и  вернулся в мир. Диего так и не пошёл. 

у себя изменил проверку
```D
func void DIA_Addon_ThiefOW_ComeOn_Info()
{
	AI_Output(other,self,"DIA_Addon_Diego_ComeOn_15_00");	//Пойдем со мной.
	if	(C_DiegoTooFar(0))
	&&	!Diego_SilvestroVariation == 1
	{
		AI_Output(self,other,"DIA_Addon_Diego_ComeOn_11_01");	//Мы идем не туда!
		AI_StopProcessInfos(self);
	}
	else
	{
		AI_Output(self,other,"DIA_Addon_Diego_ComeOn_11_02");	//Хорошо.
		AI_StopProcessInfos(self);
		Npc_ExchangeRoutine(self,"FOLLOW");
		self.aivar[AIV_PARTYMEMBER] = TRUE;
	};
};
```
теперь попробую выбрать дистанцию получше

**Comment by Kor-Angar** [comment]
> теперь попробую выбрать дистанцию получше

вот 800  https://www.youtube.com/watch?v=wXl5TQjw5Ok
```
	else if(Npc_GetDistToWP(hero,"OW_MINE3_OUT") < (800 + tolerance))
	{
		return LOC_SILVESTROMINE;
	}
```

**Comment by Kor-Angar** [comment]
но лучше 700 или 600. 
на 700 он отходит на свой чекпойнт. а вот 800 и 1000 слишком близко к чекпойнту и он стоит возле него на месте. таким образом и получается мертвая зона. 

у себя поставил на 700 завтра еще пару раз гляну.

**Comment by dosinabox** [comment]
> ```d
> !Diego_SilvestroVariation == 1
> ```

```d
Diego_SilvestroVariation != 1
```

**Comment by dosinabox** [comment]
> у себя поставил на 700 завтра еще пару раз гляну.

Исправление актуально только вместе с изменением условия диалога?

**Comment by Kor-Angar** [comment]
Условие менял для теста. Что бы подобрать дистанцию при которой Диего не попадёт в мёртвую зону. 

В итоговой версии изменил только дистанцию на 700.

**Comment by dosinabox** [comment]
Есть сохраненка?

**Comment by Kor-Angar** [comment]
не думаю что загрузит. у нас уже версии стали сильно отличаться.
[2020_11_12_D36.zip](https://github.com/dosinabox/g2nr_unofficial_update/files/5528951/2020_11_12_D36.zip)
но если загрузит там еще два сейва про вино.

**Comment by dosinabox** [comment]
> два сейва про вино.

А что там?

**Comment by Kor-Angar** [comment]
Если Диего остался ждать героя, а герой ушел по своим делам в Ярик / Хоринос, то Диего можно возвращать на исходную позицию. https://github.com/Kor-Angar/G2_NotR_Kor_Angar/commit/ad49b85f2781594b58b325dcb645b55b0e995c7a

```
	if (Diego_angekommen == FALSE)
	{
		B_StartOtherRoutine(PC_ThiefOW,"START");
	};

```

***
***

по сюжету в таком случае можно конечно и миссию закончить Диего. 

> 		AI_Output(self,other,"DIA_Addon_Diego_GoHome_11_08");	//Но не слишком задерживайся, иначе я вернусь один.

Дескать Диего сам выбрался пока герой бродил. Этот вариант слишком суровый, хочу в своей версии его реализовать, но ума как обычно не хватает :-) Диего просто пропадает :-(

```D
func void B_ENTER_OLDWORLD_Kapitel_2()
{

	// Хард вариант
	if 	Diego_XardasVariation == 1
	||	Diego_SilvestroVariation == 1
	||	Diego_GrimesVariation == 1
	||	Diego_FajethVariation == 1
	{
		if (Diego_angekommen == FALSE)
		{
			Diego_angekommen = TRUE;
			B_StartOtherRoutine(PC_ThiefOW,"PASS");
		}
		else
		{
			B_RemoveNpc(PC_ThiefOW);
		};
	};
	// Лайт вариант
	if (Diego_angekommen == FALSE)
	{
		B_StartOtherRoutine(PC_ThiefOW,"START");
	};


	if(EnterOW_Kapitel2 == FALSE)
...
```

**Comment by dosinabox** [comment]
загрузились с расширителем парсера, но вылетает при переходе в Миненталь

**Comment by Kor-Angar** [comment]
А что такое раширител парсера? 😀 А тебе зачем этот сейв нужен, ты же за 5 минут можешь с марвином воссоздать ситуацию 😀 или не можешь? 😐

**Comment by dosinabox** [comment]
> А что такое раширител парсера?

Плагин для Юниона, с которым не зависают сохраненки с удаленными переменными, константами, функциями и инстанциями. 

> или не можешь?

не всегда

**Comment by Kor-Angar** [comment]
Наверное у меня карта расширенная загрузилась?

**Comment by Kor-Angar** [comment]
Сейчас так

```D
func void B_Addon_Diego_WillWaitOutside()
{
	AI_Output(self,other,"DIA_Addon_Diego_TooFar_11_00");	//Иди первым и осмотрись. Я подожду снаружи.
};
```

```D
	else if(C_DiegoTooFar(1000) == LOC_XARDAS)
	{
		if(Diego_XardasVariation == 0)
		{
			AI_Output(self,other,"DIA_Addon_Diego_TooFar_11_14");	//Это старая башня Ксардаса. Сам он, конечно, в ней давно не появляется.
			AI_Output(self,other,"DIA_Addon_Diego_TooFar_11_15");	//Уверен, что внутри нас ждут неприятные сюрпризы.
			B_Addon_Diego_WillWaitOutside();
			Diego_XardasVariation = 1;
		}
		else
		{
			B_Addon_Diego_WillWaitOutside();
			AI_StopProcessInfos(self);
			self.aivar[AIV_PARTYMEMBER] = FALSE;
			Npc_ExchangeRoutine(self,"XARDAS");
		};
	}
```
когда герой первый раз приходит к башне,  Диего может "пропасть"  пока герой ходит по башне.


лучше вот так
```D
func void B_Addon_Diego_WillWaitOutside()
{
	AI_Output(self,other,"DIA_Addon_Diego_TooFar_11_00");	//Иди первым и осмотрись. Я подожду снаружи.
	AI_StopProcessInfos(self);
	self.aivar[AIV_PARTYMEMBER] = FALSE;
	Npc_ExchangeRoutine(self,"XARDAS");
};
```


```D
	else if(C_DiegoTooFar(1000) == LOC_XARDAS)
	{
		if(Diego_XardasVariation == 0)
		{
			AI_Output(self,other,"DIA_Addon_Diego_TooFar_11_14");	//Это старая башня Ксардаса. Сам он, конечно, в ней давно не появляется.
			AI_Output(self,other,"DIA_Addon_Diego_TooFar_11_15");	//Уверен, что внутри нас ждут неприятные сюрпризы.
			Diego_XardasVariation = 1;
		};
		B_Addon_Diego_WillWaitOutside();
	}
```

**Comment by Kor-Angar** [comment]
вот тут https://www.youtube.com/watch?v=4QRkYJxW_k0 как раз происходит диалог с Диего. я умудрённый опытом дополнительно его "паркую" через диалог "жди здесь".

**Comment by dosinabox** [comment]
У шахты Сильвестро так же:
```cpp
else if(C_DiegoTooFar(1000) == LOC_SILVESTROMINE)
{
	if(Diego_SilvestroVariation == 0)
	{
		AI_Output(self,other,"DIA_Addon_Diego_TooFar_11_19");	//Эта шахта, в которую меня поместили, когда я вернулся сюда вместе с другими заключенными.
		AI_Output(self,other,"DIA_Addon_Diego_TooFar_11_20");	//Уверен, что живых там не осталось.
		B_Addon_Diego_WillWaitOutside();
		AI_Output(self,other,"DIA_Addon_Diego_TooFar_Add_11_20");	//Но если ты будешь задерживаться, я вернусь в свое укрытие.
		AI_Output(self,other,"DIA_Addon_Diego_TooFar_Add_11_21");	//Если, конечно, ты не хочешь, чтобы я подождал тебя здесь.
		Diego_SilvestroVariation = 1;
	}
	else
	{
		B_Addon_Diego_WillWaitOutside();
		AI_StopProcessInfos(self);
		self.aivar[AIV_PARTYMEMBER] = FALSE;
		Npc_ExchangeRoutine(self,"SILVESTRO");
	};
}
```

**Comment by dosinabox** [comment]
> лучше вот так

Неправильно, не трогай B_Addon_Diego_WillWaitOutside. Надо вот так - регулируем только комментарии, но всегда меняем распорядок:
```cpp
else if(C_DiegoTooFar(1000) == LOC_XARDAS)
{
	if(Diego_XardasVariation == 0)
	{
		AI_Output(self,other,"DIA_Addon_Diego_TooFar_11_14");	//Это старая башня Ксардаса. Сам он, конечно, в ней давно не появляется.
		AI_Output(self,other,"DIA_Addon_Diego_TooFar_11_15");	//Уверен, что внутри нас ждут неприятные сюрпризы.
		Diego_XardasVariation = 1;
	};
	B_Addon_Diego_WillWaitOutside();
	AI_StopProcessInfos(self);
	self.aivar[AIV_PARTYMEMBER] = FALSE;
	Npc_ExchangeRoutine(self,"XARDAS");
}
```
```cpp
else if(C_DiegoTooFar(1000) == LOC_SILVESTROMINE)
{
	if(Diego_SilvestroVariation == 0)
	{
		AI_Output(self,other,"DIA_Addon_Diego_TooFar_11_19");	//Эта шахта, в которую меня поместили, когда я вернулся сюда вместе с другими заключенными.
		AI_Output(self,other,"DIA_Addon_Diego_TooFar_11_20");	//Уверен, что живых там не осталось.
		B_Addon_Diego_WillWaitOutside();
		AI_Output(self,other,"DIA_Addon_Diego_TooFar_Add_11_20");	//Но если ты будешь задерживаться, я вернусь в свое укрытие.
		AI_Output(self,other,"DIA_Addon_Diego_TooFar_Add_11_21");	//Если, конечно, ты не хочешь, чтобы я подождал тебя здесь.
		Diego_SilvestroVariation = 1;
	}
	else
	{
		B_Addon_Diego_WillWaitOutside();
	};
	AI_StopProcessInfos(self);
	self.aivar[AIV_PARTYMEMBER] = FALSE;
	Npc_ExchangeRoutine(self,"SILVESTRO");
}
```

**Comment by Kor-Angar** [comment]
тоже надо править. в таком состояние герой зайдёт в шахту, дойдёт до конца шахты, Вернётся, а диего уже не будет.

**Comment by Kor-Angar** [comment]
> не трогай B_Addon_Diego_WillWaitOutside. 

тогда он по сути уже и не нужен ;-)

```D
else if(C_DiegoTooFar(1000) == LOC_XARDAS)
{
	if(Diego_XardasVariation == 0)
	{
		AI_Output(self,other,"DIA_Addon_Diego_TooFar_11_14");	//Это старая башня Ксардаса. Сам он, конечно, в ней давно не появляется.
		AI_Output(self,other,"DIA_Addon_Diego_TooFar_11_15");	//Уверен, что внутри нас ждут неприятные сюрпризы.
		Diego_XardasVariation = 1;
	};
//	B_Addon_Diego_WillWaitOutside();
	AI_Output(self,other,"DIA_Addon_Diego_TooFar_11_00");	//Иди первым и осмотрись. Я подожду снаружи.
	AI_StopProcessInfos(self);
	self.aivar[AIV_PARTYMEMBER] = FALSE;
	Npc_ExchangeRoutine(self,"XARDAS");
}
```
****

для других двух шахт куда диего не хочет идти, но куда надо ГГ по основному квесту было бы хорошо дать возможность герою тут же "припарковать диего"
```D
	else if(C_DiegoTooFar(1000) == LOC_FAJETHMINE)
	{
		if(Diego_FajethVariation == 0)
		{
			AI_Output(self,other,"DIA_Addon_Diego_TooFar_11_16");	//Это шахта Фаджета.
			AI_Output(self,other,"DIA_Addon_Diego_TooFar_11_17");	//Если ты хочешь туда пойти, на меня не рассчитывай!
			Diego_FajethVariation = 1;
		}
		else
		{
			AI_Output(self,other,"DIA_Addon_Diego_TooFar_11_18");	//Я намерен держаться от этой шахты подальше!
		};
// добавить  "Подожди здесь!" или "Идём дальше!"
	}
```

я опять же  как индеец стреляный знаю что после того как диего бросил реплику тут же можно к нему обратиться и припарковать его через DIA_Addon_ThiefOW_GoHome_Info так как рядом есть точка и там диего будет есть яблоки, курить дурь и никуда не денется. но у игроков может случится что диего сказал реплику, но при этом не припарковался, герой идёт к Фаджету по квесту,  возвращается а диего уже нет.


```D
	Info_ClearChoices	(DIA_Addon_ThiefOW_TooFar);
	Info_AddChoice		(DIA_Addon_ThiefOW_TooFar,"Подожди здесь!",	DIA_Addon_ThiefOW_GoHome_Info);
	Info_AddChoice		(DIA_Addon_ThiefOW_TooFar,"Идём дальше!",	DIA_DiegoOw_EXIT_Info);
```
реплика про идём дальше по моему есть у Лареса.

***

так же у старой шахты у меня была проблема с парковкой Диего. не уверен правда что в твоей сборке. если будешь править я могу проверить.

**Comment by dosinabox** [comment]
> тогда он по сути уже и не нужен ;-)

Нужен, потому что ещё используется у шахты Сильвестро.

**Comment by dosinabox** [comment]
> для других двух шахт куда диего не хочет идти, но куда надо ГГ по основному квесту было бы хорошо дать возможность герою тут же "припарковать диего"

Диалог завершается со сменой распорядка на ожидание только у башни Ксардаса и шахты Сильвестро (после правки). У остальных мест не завершается, Диего просто комментирует обстановку и всё. Остаются такие варианты:
![image](https://user-images.githubusercontent.com/39790055/141365513-461843ff-48cd-44f5-92ff-efc63b5f2d7e.png)
"Подожди здесь!" - припарковать.
"КОНЕЦ" - продолжить маршрут.
Можно конечно добавить Info_ClearChoices, но сути это не поменяет, а только скроет "лишние" варианты. Диего можно и так припарковать везде, кроме мест, где он паркуется сам (башня Ксардаса и шахта Сильвестро).

**Comment by dosinabox** [comment]
> тогда он по сути уже и не нужен ;-)

В идеале каждая реплика должна существовать в скриптах только в одном экземпляре, чтобы не было мешанины с субтитрами и чтобы не раздувать OU.BIN. Поэтому повторные реплики засовывают в функции и переиспользуют где нужно. В случае с B_Addon_Diego_WillWaitOutside реплика DIA_Addon_Diego_TooFar_11_00 используется в трех местах.

**Comment by dosinabox** [comment]
> Можно конечно добавить Info_ClearChoices, но сути это не поменяет, а только скроет "лишние" варианты.

Точно надо? Выглядит норм)

**Comment by Kor-Angar** [comment]
Понимаешь оно когда в игре, Диего тебе говорит и есть два действия это выбор. Когда он просто сказал и выкатилось обычное большое меню, не очивидно что его можно оставить. Я к этой простой мысли приходил  годами. И на такое простое решение с Info_ClearChoices вышел тоже  несразу 😀

**Comment by Kor-Angar** [comment]
В игре оно не так очевидно 😀 легко проверить 😀

**Comment by Kor-Angar** [comment]
Точно. Он постоянно даёт какие то комментарии, а припарковать его можно в ограниченном кол-ве мест.

**Comment by Kor-Angar** [comment]
> Проверь:
> g2a_nr_scriptpatch_v27_hotfix.zip

сейв 27 версии без парсера не грузится.

***
проверка на реплику возле шахты Фаджета
```D
	else if((Npc_GetDistToWP(hero,"OW_NEWMINE_03") < (2500 + tolerance)) || (Npc_GetDistToWP(hero,"OW_NEWMINE_03_B") < (2500 + tolerance)))
	{
		return LOC_FAJETHMINE;
	}
```

не совпадает с проверкой на парковку
```D
	else if(Npc_GetDistToWP(self,"OW_NEWMINE_11") < (4000 + 1000))
	{
		AI_Output(self,other,"DIA_Addon_Diego_GoHome_11_03");	//Я подожду у шахты.
		AI_StopProcessInfos(self);
		self.aivar[AIV_PARTYMEMBER] = FALSE;
		Npc_ExchangeRoutine(self,"FAJETH");
	}
```

это наверное значит что если зайти на стоянку Фаджета через вход со стороны башни Ксарадаса, то Диего припарковать не получаться

**Reply by dosinabox** [reply]
Для меня странно, что есть целая функция C_DiegoTooFar, которая определяет местоположение Диего и возвращает точные значения, а в DIA_Addon_ThiefOW_GoHome_Info вместо нее используется что-то отдельное и проверяются какие-то другие точки. Но пока не было времени покопаться.

**Reply by dosinabox** [reply]
Т.е. вместо
```cpp
else if(Npc_GetDistToWP(self,"OW_NEWMINE_11") < (4000 + 1000))
```
должно быть
```cpp
else if(C_DiegoTooFar(1000) == LOC_FAJETHMINE)
```

**Reply by dosinabox** [reply]
На C_DiegoTooFar безболезненно меняются условия для шахт Сильвестро и Граймса, ибо одинаковые:
<img width="458" alt="Screenshot_3" src="https://user-images.githubusercontent.com/39790055/141530654-57bf8e9b-b43c-4ed1-a567-110d3779490a.png">
<img width="472" alt="Screenshot_4" src="https://user-images.githubusercontent.com/39790055/141530657-b9feeb3a-174a-4514-8ab9-d9e466c758dd.png">

**Reply by Kor-Angar** [reply]
у шахты Сильвестра и старой шахты (с рудокопом Граймс) , всё просто, так как точки одни и те же

а вот для шахты Фаджета сложнее, надо смотреть в игре все три точки, так как мне кажется нижний вход слишком далеко от места парковки диего. если просто сделать как ты написала выше Диего может пойти на место парковки через лагерь Фаджета или начать тупить об забор. надо смотреть в игре на точки. как это работает сейчас и от этого уже отталкиваться.

**Reply by dosinabox** [reply]
> а вот для шахты Фаджета сложнее, надо смотреть в игре все три точки

Все рядом, так что тоже можно перевести на единый C_DiegoTooFar:
<img width="1313" alt="Screenshot_1" src="https://user-images.githubusercontent.com/39790055/141540092-f531162c-b304-4bf6-8601-4266c5a0b37a.png">

**Comment by dosinabox** [comment]
Предлагаешь одну и туже точку/точки?

**Comment by Kor-Angar** [comment]
пока ничего не предлагаю. надо проверить в игре. потыкаться на всех точках, понять есть ли системная проблема или это частная ситуация. если ситуация частная, то и решение можно предложить частное :-)

***

я или вечером гляну или завтра.

**Comment by dosinabox** [comment]
Удалось сделать все проверки едиными через C_DiegoTooFar и по максимуму всё оптимизировать. Тестировать на этой сборке, пожалуйста:
[g2a_nr_scriptpatch_v27_hotfix.zip](https://github.com/dosinabox/g2nr_unofficial_update/files/7530632/g2a_nr_scriptpatch_v27_hotfix.zip)

**Comment by Kor-Angar** [comment]
у Диего всё вроде ок. за исключением того что пока она припаркован, он видимо не в пати и опыт не дают за убитых им монстров.

**Reply by Kor-Angar** [reply]
1 час  16 минут https://youtu.be/6rd8YIQirqw?t=4596

**Comment by Kor-Angar** [comment]
чекпойнт можно менять в расписание. что бы диего не возвращался "на старт", если герой  уйдёт за пределы видимости


```D
func void Rtn_Follow_1()
{
	var int location;
	location = C_DiegoTooFar(1000);
	
	if(location == LOC_GRIMESMINE)
	{
		TA_Follow_Player(8,0,23,0,"LOCATION_02_05");
		TA_Follow_Player(23,0,8,0,"LOCATION_02_05");
	}
	else
	{
		TA_Follow_Player(8,0,23,0,"OW_PATH_155");
		TA_Follow_Player(23,0,8,0,"OW_PATH_155");
	};
};
```

**Reply by Kor-Angar** [reply]
![20220607202356_1](https://user-images.githubusercontent.com/49514764/172445450-70dacfff-9af6-4547-9a5a-af0999946fd3.jpg)
![20220607202400_1](https://user-images.githubusercontent.com/49514764/172445457-ccd43bd8-6844-4cd8-8805-ac080ad03431.jpg)

**Reply by Kor-Angar** [reply]
это я сегодня на стрими понял неожиданно :-)

**Reply by Kor-Angar** [reply]
все точки так делать смысла нет, но вот эту точно можно, так как это бОльшая часть пути и если Диего здесь отклонится, то придётся идти сначала:-)
